### DISCUSSION TITLE
Why is Windows 11's Dynamic Dependency API not available in WinRT?

### DISCUSSION THREAD

**Discussion by roxk** [microsoft/WindowsAppSDK#3693] [Q&A]
(I know this is not the best place to ask this question, but this repository is where most of the discussion around dyndep happens and documented so I figured I might as well ask here)

In [this doc](https://learn.microsoft.com/en-us/windows/apps/desktop/modernize/framework-packages/use-the-dynamic-dependency-api#1-create-an-install-time-reference), it is written that

> 1. Create an install-time reference
> ...
> ...call one of the following functions or methods to specify a set of criteria for the MSIX package that you want to use...
> Windows 11 (C/C++): [TryCreatePackageDependency](https://learn.microsoft.com/en-us/windows/win32/api/appmodel/nf-appmodel-trycreatepackagedependency)
Windows App SDK (C/C++): [MddTryCreatePackageDependency](https://learn.microsoft.com/en-us/windows/windows-app-sdk/api/win32/msixdynamicdependency/nf-msixdynamicdependency-mddtrycreatepackagedependency)
Windows App SDK (WinRT): [PackageDependency.Create](https://learn.microsoft.com/en-us/windows/windows-app-sdk/api/winrt/microsoft.windows.applicationmodel.dynamicdependency.packagedependency.create)

Why is Windows 11's dynamic dependency not available in WinRT, but only available in plain C? It means C++ devs have to suffer from C style API (not ergonomic), and C# additionally has to write interop boilerplate. Is it an omission in the doc or?

Thanks!

**Comment by DarranRowe** [comment]
Well, I'm not in a position to answer the why question since this would require someone from Microsoft to answer. There is also nothing in Windows.ApplicationModel in the latest canary SDK (Windows SDK 25387) to show that this exists or is currently in development.
One thing to remember is that it is pretty easy to wrap this. This repo [has the implementation](https://github.com/microsoft/WindowsAppSDK/tree/main/dev/DynamicDependency/API) of the Windows App SDK version, so it can be used as a template for your own version. I think the biggest issue is figuring out how to store the install time references that use a filesystem or registry key. This could also be used as a template for a native C++ wrapper or a native C# wrapper too.
I want to reiterate that this isn't an attempt at an answer, but devs don't have to "suffer" if they don't want to.

**Comment by seven-mile** [comment]
Since WinRT is still not projected to most languages, it's the best option for Windows to export C API. Maybe in the future when WinRT is completely stable and usable, they choose to port WinRT-only API.

**Reply by roxk** [reply]
My ask isn't to remove C style API. My ask is that C style API and WinRT API should _coexist_. See e.g. app model stuff in WinRT and `appmodel.h`

**Reply by jonwis** [reply]
WinRT is stable. I wish some patterns were more easily expressed (interface-based programming, overloads with default values, native handle types without an `-Interop` interface or additional refcounting, better access to literal arrays in `IVector<T>`, more generalized generics support, cross-process WinRT activation, proxy-stub DLL generation from metadata) but we're as stable as it's going to get from a core technology POV.

This thread is asking why there isn't a `Windows.ApplicationModel.*` version of the dynamic dependency API. When we were doing design and review of this new API, we explicitly chose to put the C-style ABI into the platform SDK and to put a WinRT-based wrapper into the Windows App SDK.  That's a path I'd like for us to continue on.  Certainly, there are many APIs that work best in Windows as WinRT APIs - new experience features, new integration points for apps, updates to existing API surfaces - and we'll continue to prioritize those as WinRT in upcoming SDKs.  For this specific API, since Windows App SDK updates faster and more broadly than Windows itself, we added a `Microsoft.Windows.ApplicationModel.*` version that works "everywhere" and lights up using the platform's built-in C-style ABI `AddPackageDependency` when running on up-to-date Windows machines.

As others have noted, there's great ways to wrap the platform SDK C ABI methods for your runtime if you don't choose to use WinAppSDK.

The issue is knowing and supporting all the languages and runtimes that apps want to use. You can get a C-style export of any WinRT API using https://learn.microsoft.com/en-us/cpp/cppcx/wrl/use-winmdidl-and-midlrt-to-create-h-files-from-windows-metadata?view=msvc-170 (while the docs indicate it's for C++/CX, it's generally usable.) That gets you an "old SDK style" projection of WinRT types using the ABI COM-style vtable-with-members accessible to C consumers. See also https://blogs.windows.com/windowsdeveloper/2021/01/21/making-win32-apis-more-accessible-to-more-languages/ where the platform SDK ABI methods are converted to WinMD so they can be projected into target languages.

The other idea in this thread is something we don't have yet (and we've talked about it for a number of years) which is a single-step "WinMD -> C" projection. That'd look something like:

```c
DECLARE_HANDLE(WIN_APPMODEL_PACKAGE);
STDAPI_(void) WinAppModelPackage_Release(WIN_APPMODEL_PACKAGE package);
STDAPI WinAppModelPackage_GetCurrent(WIN_APPMODEL_PACKAGE* current);
STDAPI WinAppModelPackage_GetDescription(WIN_APPMODEL_PACKAGE package, HSTRING* description);
```

... basically "inverting" the normal C-style-COM pattern so the first parameter is the "this" and the remaining are the normal arguments.

Personally I don't find that compelling vs either "ABI COM" or "C-style-COM", but some people prefer handles and don't want to show vtables.

**Reply by seven-mile** [reply]
Oh, of course WinRT has been stable for a long time. Typo. I probably mean "the language projections of WinRT".

**Reply by DarranRowe** [reply]
@jonwis 
There is also abi.exe for the more direct WinMD -> C/C++ header creation. This can be obtained by extracting the Microsoft.Windows.AbiWinRT nuget package.
The biggest issue that I have found when accessing WinRT components through the C interface is that some IIDs aren't made available in uuid.lib. As an example, running dumpbin /symbols gives:
```
D46 00000000 SECT46D notype       External    | IID___x_Windows_CFoundation_CIGetActivationFactory
008 00000000 SECT3  notype       External     | IID___x_Windows_CFoundation_CIActivatableClassRegistration
00B 00000000 SECT4  notype       External     | IID___x_Windows_CFoundation_CIDllServerActivatableClassRegistration
00E 00000000 SECT5  notype       External     | IID___x_Windows_CFoundation_CIExeServerActivatableClassRegistration
011 00000000 SECT6  notype       External     | IID___x_Windows_CFoundation_CIExeServerRegistration
014 00000000 SECT7  notype       External     | IID___x_Windows_CFoundation_CIExtensionRegistration
01A 00000000 SECT9  notype       External     | IID___x_Windows_CFoundation_CIExtensionCatalog
01D 00000000 SECTA  notype       External     | IID___x_Windows_CFoundation_CIActivationContext
020 00000000 SECTB  notype       External     | IID___x_Windows_CFoundation_CIActivationContextFactory
023 00000000 SECTC  notype       External     | IID___x_Windows_CFoundation_CIActivatableClassRegistrationPrivate
```
for the Windows.Foundation namespace. The big thing to note is that there is nothing for the ABI there. Windows.Foundation is where interfaces such as IClosable, IStringable and runtime classes such as Uri are defined and the IIDs aren't there for the C side of things. C also doesn't let you use the __uuidof operator. This meant that I had to define these IIDs in that application.

**Reply by jonwis** [reply]
@DarranRowe thanks - you're right that we don't publish every single IID in uuid.lib. Internally we use `__uuidof()` heavily.

I'd forgotten about the abi tool from xlang, thanks for the reminder!  Does running `abi.exe` on `Windows.WinMD` metadata produce the constants you need?

I have this fever dream where the SDK contains only metadata, and everything is a set of projectors.  Even static import libraries - just generate those on the fly from APIsets, no more worrying about kernelbase vs advapi vs whatever else.  Auto-generate delayloads and presence-check methods for _any_ ABI method.  We'd_ probably cut the core install size of the SDK down by 30% or more, and we'd be more compiler-toolchain-agile. That's where we _started_ with the CS/Win32 & WinMetadata & the Microsoft.Windows.SDK.* nuget packages. Maybe we'll continue on that path.

**Reply by DarranRowe** [reply]
@jonwis 
Unfortunately, abi.exe doesn't do it either. All it generates is the same kind of external declaration, `EXTERN_C const IID IID___x_ABI_CWindows_CFoundation_CIUriRuntimeClassFactory;`, that is available in the Windows SDK headers.
This isn't the worst situation since the ABI headers also include the C++ definitions which also includes the IID. It just means that I would just define these values myself if I ever need to use the IID directly.

**Reply by DrusTheAxe** [reply]
>This thread is asking why there isn't a Windows.ApplicationModel.* version of the dynamic dependency API. When we were doing design and review of this new API, we explicitly chose to put the C-style ABI into the platform SDK and to put a WinRT-based wrapper into the Windows App SDK. That's a path I'd like for us to continue on. 

Ditto. It's been a very good path to follow and one I (continue to) recommend and advocate.

**Reply by DarranRowe** [reply]
@jonwis 
I wrote a little script to generate all of the IIDs using midl and then compiling them into a library. This ended up at around 4MiB per platform, 7 for ARM64 if you include the EC symbols. So I completely understand why these aren't shipped with the SDK and why you want a more generative version of the Windows SDK.
Since I was trying it out, I also tried this method with the Windows App SDK. I generated the IDL files using winmdidl and then used midl to generate the IID files. This actually ended up with an interesting failure. I was getting errors along the lines of CrossSlidingState_Started already defined. Looking at the imports a little, microsoft.ui.input.idl imports windows.ui.core.idl, which in turn imports windows.ui.input.idl. Since midl doesn't seem to mangle the names of enumeration members to differentiate the scopes, this does end up with an error because the enumerations in Windows.UI.Input are mirrored in Microsoft.UI.Input. Windows.UI.Input and Microsoft.UI.Input aren't the only affected namespaces. That was just the first set of errors I saw. So unfortunately this means that using midl isn't a good general means of generating the IIDs for the C interfaces.
A little tool for reading the metadata and generating these IIDs would be useful for anyone wanting to use the C interfaces.

**Reply by jonwis** [reply]
Neato! Sounds like a fine tool to share if you're willing.

We do rely on namespaces heavily, especially where `Microsoft.UI.*` aims to replace/supplant `Windows.UI.*` types. For languages that support scope aliases it's not super hard to deal with - `wui::CrossSlidingState` vs `mui::CrossSlidingState` - but for languages without, life gets hard. Note how MIDL generates _massively_ long interface names with the full namespace associated with them.  Maybe in your tool, apply common shortenings or single-characters? Not super nice, but `M_UI_CrossSliding` vs `WIN_UI_CrossSliding` wouldn't be _so_ horrible?

**Comment by DrusTheAxe** [comment]
>I know this is not the best place to ask this question, but

It's actually a pretty good place for it :-)

>Why is Windows 11's dynamic dependency not available in WinRT

Not available _yet_ but soon => [DynamicDependencies: Use Win11 APIs if available #3704](https://github.com/microsoft/WindowsAppSDK/pull/3704)

There's some work remaining but given the interest I've shared some of the more stable bits. I have a more unstable build locally I'm fiddling with e.g. tests are blocked because the bootstrapper's not savvy yet. Working through the details.

The goal is for WinAppSDK's Dynamic Dependency API to delegate to the OS' DynDep API if available, else use its own implementation to equivalent end (aka polyfill). This will happen in the WinAppSDK C API so all calls of WinAppSDK's C and WinRT DynDep API will benefit (including those using the C# projection).

This has additional benefits, e.g. the DDLM packages are unnecessary (when the Win11 API's available). They're a workaround to make Windows DoTheRightThing(TM) e.g. not remove a package while dynamically in use. But Windows handles that now (when using the Win11 API). Likewise, WinAppSDK's DynDep only supporting unpackaged apps is a limitation of its implementation; we can lift that restriction using the Win11 API. You can see hints of this already in the draft PR, with more to come.

**Reply by sylveon** [reply]
Does that work address issues like #3242 while using the WASDK version of dyn deps? Otherwise this would cause inconsistent behavior downlevel (which kinda defeats the point of a polyfill). My current use case for dynamic dependencies is primarily to import WinUI 2, so at the moment I cannot do any sort of polyfilling.

**Comment by DrusTheAxe** [comment]
> Does that work address issues like #3242 while using the WASDK version of dyn deps?

Depends on the root cause. I'll need to hunt up some WinUI and MRT experts to say more.

>Otherwise this would cause inconsistent behavior downlevel (which kinda defeats the point of a polyfill). My current use case for dynamic dependencies is primarily to import WinUI 2, so at the moment I cannot do any sort of polyfilling.

Does your issue with WinUI2 + WinAppSDK DynDep **+ Win11**? Your issue doesn't say what OS you're running on when it fails, only that WinUI2 + **Win11** DynDep + Win11 works.

**Reply by sylveon** [reply]
> Does your issue with WinUI2 + WinAppSDK DynDep + Win11?

I believe so yes.
